File: 700/745 - Prefix and Suffix Search.java
import java.util.Map;
import java.util.HashMap;

class WordFilter {
    
    Map<String, Integer> map;

    public WordFilter(String[] words) {
        map = new HashMap<>();
        StringBuilder sb = new StringBuilder();
        
        for (int i = 0; i < words.length; ++i) {
            String word = words[i];
            int len = word.length();
            
            for (int j = 0; j <= 10 && j <= len; ++j) {
                for (int k = 0; k <= 10 && k <= len; ++k) {
                    sb.append(word.substring(0, j));
                    sb.append('#');
                    sb.append(word.substring(len - k));
                    
                    map.put(sb.toString(), i);
                        
                    sb.setLength(0);
                }
            }
        }
    }
    
    public int f(String prefix, String suffix) {
        return map.getOrDefault(prefix + "#" + suffix, -1);
    }
}

/**
 * Your WordFilter object will be instantiated and called as such:
 * WordFilter obj = new WordFilter(words);
 * int param_1 = obj.f(prefix,suffix);
 */

File: csharp/745.prefix-and-suffix-search.cs
/*
 * @lc app=leetcode id=745 lang=csharp
 *
 * [745] Prefix and Suffix Search
 */

using System.Collections.Generic;

// @lc code=start
public class WordFilter {

    private IDictionary<string, int> Map;

    public WordFilter(string[] words) {
        Map = new Dictionary<string, int>();
        for (int i = 0; i < words.Length; ++i) {
            var word = words[i];
            int n = word.Length;
            string[] prefixes = new string[n + 1], suffixes = new string[n + 1];
            for (int j = 0; j <= n; ++j) {
                prefixes[j] = word.Substring(0, j);
                suffixes[j] = word.Substring(n - j);
            }
            foreach (var prefix in prefixes) {
                foreach (var suffix in suffixes) {
                    Map[$"{prefix}-{suffix}"] = i;
                }
            }
        }
    }
    
    public int F(string prefix, string suffix) {
        if (Map.TryGetValue($"{prefix}-{suffix}", out int ret)) {
            return ret;
        }
        return -1;
    }
}

/**
 * Your WordFilter object will be instantiated and called as such:
 * WordFilter obj = new WordFilter(words);
 * int param_1 = obj.F(prefix,suffix);
 */
// @lc code=end

